destination/config.go

```diff
@@ -0,0 +1,190 @@
+// Copyright © 2022 Meroxa, Inc. and Miquido
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package destination
+
+import (
+	"fmt"
+	"strconv"
+	"strings"
+
+	"github.com/miquido/conduit-connector-elasticsearch/internal/elasticsearch"
+)
+
+const (
+	ConfigKeyVersion                = "version"
+	ConfigKeyHost                   = "host"
+	ConfigKeyUsername               = "username"
+	ConfigKeyPassword               = "password"
+	ConfigKeyCloudID                = "cloudId"
+	ConfigKeyAPIKey                 = "apiKey"
+	ConfigKeyServiceToken           = "serviceToken"
+	ConfigKeyCertificateFingerprint = "certificateFingerprint"
+	ConfigKeyIndex                  = "index"
+	ConfigKeyType                   = "type"
+	ConfigKeyBulkSize               = "bulkSize"
+	ConfigKeyRetries                = "retries"
+)
+
+type Config struct {
+	Version                elasticsearch.Version
+	Host                   string
+	Username               string
+	Password               string
+	CloudID                string
+	APIKey                 string
+	ServiceToken           string
+	CertificateFingerprint string
+	Index                  string
+	Type                   string
+	BulkSize               uint64
+	Retries                uint8
+}
+
+func (c Config) GetHost() string {
+	return c.Host
+}
+
+func (c Config) GetUsername() string {
+	return c.Username
+}
+
+func (c Config) GetPassword() string {
+	return c.Password
+}
+
+func (c Config) GetCloudID() string {
+	return c.CloudID
+}
+
+func (c Config) GetAPIKey() string {
+	return c.APIKey
+}
+
+func (c Config) GetServiceToken() string {
+	return c.ServiceToken
+}
+
+func (c Config) GetCertificateFingerprint() string {
+	return c.CertificateFingerprint
+}
+
+func (c Config) GetIndex() string {
+	return c.Index
+}
+
+func (c Config) GetType() string {
+	return c.Type
+}
+
+func ParseConfig(cfgRaw map[string]string) (_ Config, err error) {
+	cfg := Config{
+		Version:                cfgRaw[ConfigKeyVersion],
```

Comment:
We do, since it is not always available immediately. The server can require the client to be authorized first. In such case, I would need to use some version of the client, see if it connects, get the version and then use the proper version of the client.

Reply:
Gotcha. However, the request to localhost:9200 is like any other request. It can fail only if the config is missing credentials, in which case the connector is not able to work anyway. In other words, we have two possibilties:
1. ES server requires auth, but the connector config doesn't have it: We will ping localhost:9200, get a 401 code back, and will let the user know that the config is invalid (missing creds).
2. ES doesn't require auth, the connector config doesn't have it: all good
3. ES requires auth, connector config has it: pinging the host to get the version will work.

Reply:
Assuming e.g. ES SDK v8 (and future versions) can authorize ES server v5 (or any previous version in general). A similar thing happened to MySQL connector for PHP after MySQL 8.0 was released, and they changed the default auth plugin (https://stackoverflow.com/questions/49083573/php-7-2-2-mysql-8-0-pdo-gives-authentication-method-unknown-to-the-client-ca) which lead to kind of false-positive where user+password are correct, but additional configuration parameter is required.

However, I think we are good to go with this. I see it currently as a rare case, since ES is pretty much robust with their changes (they either do not change already well working features or announce changes early enough). I've created an issue for that https://github.com/miquido/conduit-connector-elasticsearch/issues/4.